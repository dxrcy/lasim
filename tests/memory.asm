.ORIG x3000
    lea r1, A
    REG
A
    lea r1, A
    REG

    ld r1, X
    REG

    not r1, r1
    st r1, X
    REG

    ld r1, X
    REG

    ld r1, InstrValue
    REG
    st r1, InstrLocation

InstrLocation
    .FILL 0xd000
    REG

    ld r1, Addr
    REG
    and r2, r2, #0
    add r2, r2, x000e
    str r2, r1, #0

    ld r1, Addr
    REG
    ldr r1, r1, #0
    REG

    and r1, r1, #0
    REG
    ldi r1, Addr
    REG

    add r1, r1, #-5
    REG
    sti r1, Addr

    and r1, r1, #0
    REG
    ldi r1, Addr
    REG

    HALT

X           .FILL xbeef
InstrValue  .FILL x1265 ; add r1, r1, #5
Addr        .FILL x3100

.END
