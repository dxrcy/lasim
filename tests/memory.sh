#!/bin/sh

source "$(dirname $0)/shared.sh"

asm_file="$tests/memory.asm"
obj_file="$out/memory.obj"
output_actual_file="$out/memory.actual"
output_expected_file="$tests/memory.expected"

extract_reg() {
    sed -n 's/ *. *r1 *\([^ ]*\).*/\1/p'
}

lasim -a "$asm_file" -o "$obj_file"
lasim -x "$obj_file" | extract_reg > "$output_actual_file"

diff "$output_expected_file" "$output_actual_file"
report_status $?


