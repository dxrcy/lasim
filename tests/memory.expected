0x3002
0x3002
0xbeef
0x4110
0x4110
0x1265
0x126a
0x3100
0x3100
0x000e
0x0000
0x000e
0x0009
0x0000
0x0009
